File: PR224-ScrollAwareFab/app/src/main/java/es/iessaladillo/pedrojoya/pr224/base/ScrollAwareFabBehavior.java
package es.iessaladillo.pedrojoya.pr224.base;

import android.animation.Animator;
import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.view.ViewCompat;

public class ScrollAwareFabBehavior extends FloatingActionButton.Behavior {

    private static final long ANIMATION_DURATION = 200;

    private boolean isHidding;

    public ScrollAwareFabBehavior(Context context, AttributeSet attrs) {
        super();
    }

    @Override
    public boolean onStartNestedScroll(@NonNull final CoordinatorLayout coordinatorLayout,
            @NonNull final FloatingActionButton child, @NonNull final View directTargetChild,
            @NonNull final View target, final int nestedScrollAxes, int type) {
        // Sólo estamos interesados en el scroll vertical.
        return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL ||
                super.onStartNestedScroll(coordinatorLayout, child,
                        directTargetChild, target,
                        nestedScrollAxes, type);
    }

    @Override
    public void onNestedScroll(@NonNull final CoordinatorLayout coordinatorLayout,
            @NonNull final FloatingActionButton child, @NonNull final View target, final int dxConsumed,
            final int dyConsumed, final int dxUnconsumed, final int dyUnconsumed,
            int type) {
        super.onNestedScroll(coordinatorLayout, child, target,
                dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed, type);
        // Si se hace scroll hacia arriba y no está oculto, se oculta.
        if (dyConsumed > 0 && !isHidding) {
            child.animate().scaleX(0).scaleY(0).rotation(-90)
                    .setDuration(ANIMATION_DURATION).setListener(
                    new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animation) {
                            isHidding = true;
                        }

                        @Override
                        public void onAnimationEnd(Animator animation) {
                        }

                        @Override
                        public void onAnimationCancel(Animator animation) {

                        }

                        @Override
                        public void onAnimationRepeat(Animator animation) {

                        }
                    });

        } else if (dyConsumed < 0 && isHidding) {
            // Si se hace scroll hacia abajo y está oculto, se hace visible.
            child.animate().scaleX(1).scaleY(1).rotation(0)
                    .setDuration(ANIMATION_DURATION).setListener(
                    new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animation) {
                            isHidding = false;
                        }

                        @Override
                        public void onAnimationEnd(Animator animation) {
                        }

                        @Override
                        public void onAnimationCancel(Animator animation) {

                        }

                        @Override
                        public void onAnimationRepeat(Animator animation) {

                        }
                    });
        }
    }

}

File: PR224-ScrollAwareFab/app/src/main/java/es/iessaladillo/pedrojoya/pr224/base/ScrollAwareFabBehavior.kt
package es.iessaladillo.pedrojoya.pr224.base

import android.animation.Animator
import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.ViewCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton

private const val ANIMATION_DURATION: Long = 200

@Suppress("unused")
class ScrollAwareFabBehavior(context: Context, attrs: AttributeSet) : FloatingActionButton.Behavior() {

    private var isHidden: Boolean = false

    override fun onStartNestedScroll(coordinatorLayout: CoordinatorLayout,
                                     child: FloatingActionButton, directTargetChild: View, target: View,
                                     nestedScrollAxes: Int, type: Int): Boolean {
        return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL || super.onStartNestedScroll(
                coordinatorLayout, child, directTargetChild, target, nestedScrollAxes, type)
    }


    override fun onNestedScroll(coordinatorLayout: CoordinatorLayout,
                                child: FloatingActionButton, target: View, dxConsumed: Int,
                                dyConsumed: Int, dxUnconsumed: Int, dyUnconsumed: Int, type: Int) {
        super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed, dxUnconsumed,
                dyUnconsumed, type)
        if (dyConsumed > 0 && !isHidden) {
            child.animate().scaleX(0f).scaleY(0f).rotation(-90f).setDuration(ANIMATION_DURATION)
                    .setListener(
                            object : Animator.AnimatorListener {
                                override fun onAnimationStart(animation: Animator) {
                                    isHidden = true
                                }

                                override fun onAnimationEnd(animation: Animator) {}

                                override fun onAnimationCancel(animation: Animator) {

                                }

                                override fun onAnimationRepeat(animation: Animator) {

                                }
                            })

        } else if (dyConsumed < 0 && isHidden) {
            child.animate().scaleX(1f).scaleY(1f).rotation(0f).setDuration(ANIMATION_DURATION)
                    .setListener(
                            object : Animator.AnimatorListener {
                                override fun onAnimationStart(animation: Animator) {
                                    isHidden = false
                                }

                                override fun onAnimationEnd(animation: Animator) {}

                                override fun onAnimationCancel(animation: Animator) {

                                }

                                override fun onAnimationRepeat(animation: Animator) {

                                }
                            })
        }
    }

}